core/embed/rust/src/ui/model_mercury/component/button.rs

```diff
@@ -527,44 +529,49 @@ impl IconText {
     ) {
         let width = self.text.map(|t| style.font.text_width(t));
 
-        let mut use_icon = false;
-        let mut use_text = false;
-
-        let mut icon_pos = Point::new(
-            area.top_left().x + ((Self::ICON_SPACE + Self::ICON_MARGIN) / 2),
-            area.center().y,
-        );
-        let mut text_pos = area.left_center() + baseline_offset;
-
         if area.width() > (Self::ICON_SPACE + Self::TEXT_MARGIN + width) {
-            //display both icon and text
-            text_pos = Point::new(area.top_left().x + Self::ICON_SPACE, text_pos.y);
-            use_text = true;
-            use_icon = true;
-        } else if area.width() > (width + Self::TEXT_MARGIN) {
-            use_text = true;
-        } else {
-            //if we can't fit the text, retreat to centering the icon
-            icon_pos = area.center();
-            use_icon = true;
-        }
-
-        if use_text {
             self.text.map(|t| {
+                let mut text_pos = area.left_center() + baseline_offset;
+                text_pos = Point::new(area.top_left().x + Self::ICON_SPACE, text_pos.y);
                 shape::Text::new(text_pos, t)
                     .with_font(style.font)
                     .with_fg(style.text_color)
                     .with_alpha(alpha)
                     .render(target)
             });
+        } else {
+            self.text.map(|t| {
+                let mut rects: Vec<Rect, 2> = Vec::new();
+                let _ = rects.push(Rect::new(area.top_left(), area.right_center()));
+                let _ = rects.push(Rect::new(area.left_center(), area.bottom_right()));
+                let mut texts: Vec<&str, 2> = Vec::new();
+                let (t1, t2) = split_two_lines(
+                    t,
+                    style.font,
+                    area.width() - Self::ICON_SPACE - Self::TEXT_MARGIN,
+                );
+                let _ = texts.push(t1);
+                let _ = texts.push(t2);
```

Comment:
you are uselessly making two vecs when (a) a fixed size arrays would be much more appropriate, but (b) even better, just make a function `fn show_text(text: &str, rect: Rect)` here and call it twice with explicit arguments

also this code is not handling the situation where either t1 or t2 is empty

in fact you should probably try "split two lines" instead of the width calculation on L530:

* t1.is_empty() || t2.is_empty(), i.e., either the text fits, or it can't be split? then render `t` on single line
* else render two lines

Reply:
Indeed, looks better with a function (a closure, actually, since I need extra context). https://github.com/trezor/trezor-firmware/pull/4226/commits/92302c1513c3aaf5c4f3634de07b9d3e65da183b#diff-70f3634936c5c39943c5bcd657c91fd5d0133e243c64994798228e192627d774R530